### PR TITLE
Reliable unlinking/case 1

### PR DESCRIPTION
**Pull request by andpodob**
This PR:
* adds pad_info to `%Endpoint` structure
* adds handling the case when one of the sides of the link is not alive and the other side wants to connect via static pad
* adds the test of handling that case

### CODE REVIEW

File: lib/membrane/core/parent/child_life_controller/link_handler.ex [hunk 1]
```diff
@@ -266,8 +266,22 @@ defmodule Membrane.Core.Parent.ChildLifeController.LinkHandler do
...
-      update_in(state, [:links], &[%Link{from: from, to: to} | &1])
+      from_availability = Pad.availability_mode(from.pad_info.availability)
+      to_availability = Pad.availability_mode(to.pad_info.availability)
+
+      case {Message.call(from.pid, :handle_link, [:output, from, to, %{initiator: :parent}]),
+            from_availability, to_availability} do
+        {:ok, _form_availability, _to_availability} ->
+          update_in(state, [:links], &[%Link{from: from, to: to} | &1])
+
+        {{:error, {:call_failure, _reason}}, :static, _to_availability} ->
```
Comment (mat-hek) [inline lib/membrane/core/parent/child_life_controller/link_handler.ex:277]: Shouldn't we kill `to.pid` when `to_availability` is `:static`?

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: test/membrane/integration/linking_test.exs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    spec_2 = %Membrane.ParentSpec{
+      children: [
+        sink: Testing.Sink
+      ],
+      crash_group: {:group_2, :temporary}
+    }
+
+    links_spec = %Membrane.ParentSpec{
+      links: [
+        link(:source) |> to(:sink)
```
Comment (mat-hek) [inline test/membrane/integration/linking_test.exs:43]: a test when the source has a dynamic output would be useful

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/core/parent/child_life_controller/link_handler.ex (hunk 1 above)
Comment (andpodob) [reply lib/membrane/core/parent/child_life_controller/link_handler.ex]: You're right.

File: lib/membrane/core/element/pad_controller.ex [hunk 3]
```diff
@@ -74,33 +74,39 @@ defmodule Membrane.Core.Element.PadController do
...
+            other_info,
+            link_metadata,
+            state
+          )
+
+        maybe_handle_pad_added(endpoint.pad_ref, state)
 
-    maybe_handle_pad_added(endpoint.pad_ref, state)
+      {:error, {:call_failure, _reason} = reason} ->
+        {{:error, {:neighbor_not_alive, reason}}, state}
```
[the remark this answers is not in the document]
Comment (andpodob) [reply lib/membrane/core/element/pad_controller.ex]: Done.

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: test/membrane/integration/linking_test.exs [hunk 4]
```diff
@@ -0,0 +1,163 @@
+defmodule Membrane.Integration.LinkingTest do
+  use ExUnit.Case
+
+  import Membrane.Testing.Assertions
+  import Membrane.ParentSpec
+
+  alias Membrane.{Buffer, Testing}
+
+  defmodule DynamicSource do
```
[the remark this answers is not in the document]
Comment (andpodob) [reply test/membrane/integration/linking_test.exs:9]: To make it really generic I would add a logic that sends all buffers passed through the option `output` to all connected pads. Now let's say you connect two Testing.Sinks, it is unpredictable which buffers will end up in which Testing.Sink, so it effectively makes it impossible to use `assert_sink_buffer`.

[the remark this answers is not in the document]
Comment (andpodob) [reply test/membrane/integration/linking_test.exs]: done.

File: test/membrane/integration/linking_test.exs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+    @impl true
+    def handle_init(opts) do
+      {:ok, Map.from_struct(opts)}
+    end
+
+    @impl true
+    def handle_prepared_to_playing(_ctx, state) do
+      {{:ok, caps: {:output, state.caps}}, state}
+    end
```
[the remark this answers is not in the document]
Comment (andpodob) [reply test/membrane/integration/linking_test.exs]: done.

File: test/membrane/integration/linking_test.exs [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      if length(state.output) > 0 do
+        [payload | rest] = state.output
+
+        {{:ok,
+          [
+            {:buffer, {pad, %Buffer{payload: payload}}},
+            {:redemand, pad}
+          ]}, %{state | output: rest}}
+      else
+        {{:ok, [end_of_stream: :output]}, state}
```
[the remark this answers is not in the document]
Comment (andpodob) [reply test/membrane/integration/linking_test.exs]: done.

File: lib/membrane/core/element/pad_controller.ex [hunk 7]
```diff
@@ -148,12 +154,16 @@ defmodule Membrane.Core.Element.PadController do
   @spec handle_unlink(Pad.ref_t(), Core.Element.State.t()) ::
           Type.stateful_try_t(Core.Element.State.t())
   def handle_unlink(pad_ref, state) do
-    with {:ok, state} <- flush_playback_buffer(pad_ref, state),
+    with :ok <- Membrane.Core.Child.PadModel.assert_instance(state, pad_ref),
```
[the remark this answers is not in the document]
Comment (andpodob) [reply lib/membrane/core/element/pad_controller.ex]: done.

**Review by mat-hek — DISMISSED** [review]

**Review by mat-hek — APPROVED** [review]

**Pull request merged**